### ISSUE TITLE
amplify codegen doesn't properly handle AppSync graphql array args, or required array args

### ISSUE THREAD

**Issue by natebushman** [aws-amplify/amplify-cli#377]
Using:

"@aws-amplify/cli": "^0.1.29"
"aws-amplify": "^1.1.5",

Steps:

When the Amplify CLI is used (with a command like "amplify update codegen") to auto-generate mutations.js and/or queries.js and/or subscriptions.js for an [*existing* AppSync API (where the API was added to the project using "amplify add codegen --apiId <the API ID>")](https://github.com/aws-amplify/amplify-cli/blob/master/native_guide.md), the auto-generated code is incorrect in cases where the mutation/query/subscription is defined with **arguments that are of type array**.  Additionally, the auto-generated string for the mutation/query/subscription is also incorrect if the argument a non-null/required array.

For instance, in the AppSync web console, we have this mutation:

`addUsersToList(listId: String!, users: [UserBasicInput!]): [UserBasic]`

The auto-generated mutations.js code (output from "amplify update codegen") looks like this:

```
export const addUsersToList = `mutation AddUsersToList($listId: String!, $users: UserBasicInput) {
  addUsersToList(listId: $listId, users: $users) {
    userId
    userName
  }
}
`
```

Notice how the above auto-generated string is not enclosing UserBasicInput in brackets, nor is it using '!' to indicate that the parameter is required/non-null.  

Rather than the above auto-generated string, we would expect it to look more like this:

```
export const addUsersToList = `mutation AddUsersToList($listId: String!, $users: [UserBasicInput!]) {
        addUsersToList(listId: $listId, users: $users) {
            userId
            userName
        }
}`
```

**The result is that when the auto-generate addUsersToList is passed into the AWS.graphql() function, a failure will occur due to the type mismatch.  The error will look like this:

"message": "Validation error of type VariableTypeMismatch: Variable type 'UserBasicInput' doesn't match expected type '[UserBasicInput!]' @ 'addUsersToList'"**

The schema.json does indeed identify the arg as an array and also that it is required.  Here's the relevant JSON object for the definition of this mutation.  Notice how the "users" arg is of kind "NON_NULL" and "LIST" (a required array):

```
{
          "name" : "addUsersToList",
          "description" : null,
          "args" : [ {
            "name" : "listId",
            "description" : null,
            "type" : {
              "kind" : "NON_NULL",
              "name" : null,
              "ofType" : {
                "kind" : "SCALAR",
                "name" : "String",
                "ofType" : null
              }
            },
            "defaultValue" : null
          }, {
            "name" : "users",
            "description" : null,
            "type" : {
              "kind" : "LIST",
              "name" : null,
              "ofType" : {
                "kind" : "NON_NULL",
                "name" : null,
                "ofType" : {
                  "kind" : "INPUT_OBJECT",
                  "name" : "UserBasicInput",
                  "ofType" : null
                }
              }
            },
            "defaultValue" : null
          } ],
          "type" : {
            "kind" : "LIST",
            "name" : null,
            "ofType" : {
              "kind" : "OBJECT",
              "name" : "UserBasic",
              "ofType" : null
            }
          },
          "isDeprecated" : false,
          "deprecationReason" : null
        }
```

As a side-note, I'm really glad this auto-generation feature was added to the amplify CLI.  It's handy.  Just needs some polishing.  Thanks guys!

**Comment by natebushman** [comment]
My guess is that this has something to do with missing fields in GQLTemplateArgDeclaration in amplify-cli/packages/amplify-graphql-docs-generator/src/generator/types.ts 

When the graphql is being generated, getArgs() is called and is used with GQLTemplateArgDeclaration.  getArgs() does indeed return whether the arg is an array (isList) and if it is required (isListRequired), however GQLTemplateArgDeclaration doesn't have any fields for these values and so perhaps they're being ignored.

Here's the generateOperation() call where getArgs() is used by GQLTemplateArgDeclaration, in amplify-cli/packages/amplify-graphql-docs-generator/src/generator/generateOperation.ts:

```
export default function generateOperation(
  operation: GraphQLField<any, any>,
  schema: GraphQLSchema,
  maxDepth: number = 3
): GQLTemplateGenericOp {
  const args: Array<GQLTemplateArgDeclaration> = getArgs(operation.args)
  const body: GQLTemplateOpBody = getBody(operation, schema, maxDepth)
  return {
    args,
    body,
  }
}
```

Here's getArgs (from amplify-cli/packages/amplify-graphql-docs-generator/src/generator/getArgs.ts).  Notice how it returns isList and isListRequired:

```
export default function getArgs(args: GraphQLArgument[]): Array<GQLTemplateArgDeclaration> {
  const argMaps = args.map((arg: GraphQLArgument) => {
    return {
      name: arg.name,
      type: getType(arg.type).name,
      isRequired: isRequired(arg.type),
      isList: isList(arg.type),
      isListRequired: isRequiredList(arg.type),
      defaultValue: arg.defaultValue,
    }
  })
  return argMaps
}
```

However the GQLTemplateArgDeclaration type doesn't include any corresponding fields for isList and isListRequired.   It's in: amplify/amplify-cli/packages/amplify-graphql-docs-generator/src/generator/types.ts and looks like this:

```
export type GQLTemplateArgDeclaration = {
  name: string
  type: string
  isRequired: boolean
  defaultValue: string | null
}
```

**Comment by natebushman** [comment]
So as an experiment I cloned the amplify-cli repo and then modified GQLTemplateArgDeclaration so that it looks like this (I added the missing isList and isListRequired fields):

```
export type GQLTemplateArgDeclaration = {
  name: string
  type: string
  isRequired: boolean
  isList: boolean
  isListRequired: boolean
  defaultValue: string | null
}
```

I then rebuilt amplify and ran the "amplify update codegen" command and the output is definitely more correct, although not entirely correct.  The generated output for the aforementioned addUsersToList mutation now looks like this (below).  Notice how [UserBasicInput] is now enclosed in brackets (it wasn't before), which is correct, however as this is a required type, in order for it to be perfectly correct it should include the '!' character, like this:  [UserBasicInput!]

```
export const addUsersToList = `mutation AddUsersToList($listId: String!, $users: [UserBasicInput]) {
  addUsersToList(listId: $listId, users: $users) {
    userId
    userName
  }
}
`;
```

**Comment by yuth** [comment]
@natebushman Thank you for reporting this issue. I have fixed this in #336. The change is pending release.  But you could get the fix by installing canary version from NPM. I am closing this issue, but feel free to re-open it, if you think the issue is not fixed.
